src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -1323,3 +1350,205 @@ fn args_required_const(tcx: TyCtxt, def_id: DefId) -> Option<FxHashSet<usize>> {
     }
     Some(ret)
 }
+
+// `root` is a const fn, check whether it can be promoted
+struct PromotableConstFn<'a, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    remaining: Vec<DefId>,
+    seen: FxHashSet<DefId>,
+}
+
+impl<'a, 'tcx> PromotableConstFn<'a, 'tcx> {
+    fn is_promotable(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> bool {
```

Comment:
Maybe this should be a query?

Reply:
I tried that originally. I just called `mir_qualif_map` on the `def_id`, and immediately got cycles for

```rust
fn a() { b() }
fn b() { a() }
```

Even if cycle errors are fine here for now, in the future with conditions we really don't want this anymore.

Reply:
Ah, okay. cc @nikomatsakis another case where we'd like queries for cycle-aware analyses. Not sure what the best approach is though. I guess recomputation shouldn't be too expensive, for now.

Reply:
I actually made it in a way that it only recomputes if the item is local and the mir of the item has not already been stolen. Kind of a crappy way to detect "no cycle here"-guarantees, but it works :laughing: